File: AIQuotaBar/Services/MiniMaxActivityDetector.swift
import Foundation

/// Detects active MiniMax coding tasks from both local runtimes:
///
/// - **MiniMax CLI** sessions under `~/.minimax/v2/sessions/<y>/<m>/<d>/`.
///   Read-only: activity is the newest file mtime inside a session directory.
///   `manifest.json` is only used for the canonical session ID — live
///   measurements show it can lag minutes behind actual transcript writes.
/// - **Background tasks** under `~/.minimax/background-tasks/bg_<uuid>/`
///   (how the MiniMax Code desktop app runs tasks). Activity is the mtime of
///   `output.log` / `summary.txt` / the task directory itself.
///
/// Message bodies and tool output are never read.
final class MiniMaxActivityDetector: ProviderLocalActivityProviding,
    @unchecked Sendable
{
    static let defaultFreshnessWindow: TimeInterval = 120

    /// How many most-recent day folders (by `YYYY/MM/DD` name) are scanned
    /// each pass. Sessions are created under the day they start, so scanning
    /// the newest two covers anything still working, including across
    /// midnight, without walking the whole history.
    static let scannedDayFolderLimit = 2

    let sessionsRootURL: URL
    let backgroundTasksRootURL: URL
    let freshnessWindow: TimeInterval

    init(
        sessionsRootURL: URL = MiniMaxActivityDetector.defaultSessionsRootURL(),
        backgroundTasksRootURL: URL = MiniMaxActivityDetector
            .defaultBackgroundTasksRootURL(),
        freshnessWindow: TimeInterval = MiniMaxActivityDetector
            .defaultFreshnessWindow
    ) {
        self.sessionsRootURL = sessionsRootURL
        self.backgroundTasksRootURL = backgroundTasksRootURL
        self.freshnessWindow = freshnessWindow
    }

    static func defaultHomeURL(
        environment: [String: String] = ProcessInfo.processInfo.environment,
        homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
    ) -> URL {
        if let configured = environment["MINIMAX_HOME"],
           !configured.isEmpty {
            return URL(fileURLWithPath: configured, isDirectory: true)
        }
        return homeDirectory.appendingPathComponent(
            ".minimax",
            isDirectory: true)
    }

    static func defaultSessionsRootURL(
        environment: [String: String] = ProcessInfo.processInfo.environment,
        homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
    ) -> URL {
        defaultHomeURL(environment: environment, homeDirectory: homeDirectory)
            .appendingPathComponent("v2", isDirectory: true)
            .appendingPathComponent("sessions", isDirectory: true)
    }

    static func defaultBackgroundTasksRootURL(
        environment: [String: String] = ProcessInfo.processInfo.environment,
        homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
    ) -> URL {
        defaultHomeURL(environment: environment, homeDirectory: homeDirectory)
            .appendingPathComponent("background-tasks", isDirectory: true)
    }

    /// Local presence check for task-protection eligibility.
    static func isClientInstalled(
        environment: [String: String] = ProcessInfo.processInfo.environment,
        fileManager: FileManager = .default
    ) -> Bool {
        let home = defaultHomeURL(environment: environment)
        return fileManager.fileExists(atPath: home.path)
    }

    func snapshot() async -> ProviderLocalActivitySnapshot {
        await Task.detached(priority: .utility) { [self] in
            detectSnapshot(now: Date())
        }.value
    }

    func detectSnapshot(now: Date) -> ProviderLocalActivitySnapshot {
        let cutoff = now.addingTimeInterval(-freshnessWindow)
        var activeSessionIDs = Set<String>()
        var lastEventBySession: [String: Date] = [:]
        var lastEventAt: Date?

        func record(_ prefixedID: String, _ activityAt: Date) {
            activeSessionIDs.insert(prefixedID)
            lastEventBySession[prefixedID] = activityAt
            if lastEventAt == nil || activityAt > lastEventAt! {
                lastEventAt = activityAt
            }
        }

        for sessionDirectory in recentSessionDirectories() {
            guard let activityAt = newestFileModification(
                in: sessionDirectory) else {
                continue
            }
            guard activityAt >= cutoff, activityAt <= now else { continue }
            let sessionID = readSessionID(at: sessionDirectory)
                ?? sessionDirectory.lastPathComponent
            record("minimax:cli:\(sessionID)", activityAt)
        }

        for taskDirectory in backgroundTaskDirectories() {
            guard let activityAt = newestBackgroundTaskModification(
                in: taskDirectory.url,
                directoryModifiedAt: taskDirectory.modifiedAt) else {
                continue
            }
            guard activityAt >= cutoff, activityAt <= now else { continue }
            record("minimax:bg:\(taskDirectory.url.lastPathComponent)", activityAt)
        }

        return ProviderLocalActivitySnapshot(
            activeSessionIDs: activeSessionIDs,
            lastEventAt: lastEventAt,
            lastEventBySession: lastEventBySession)
    }

    // MARK: - CLI sessions

    private func recentSessionDirectories() -> [URL] {
        let dayFolders = sortedDayFolderPaths().prefix(
            Self.scannedDayFolderLimit)
        var directories: [URL] = []
        for dayFolder in dayFolders {
            let dayURL = sessionsRootURL
                .appendingPathComponent(dayFolder, isDirectory: true)
            guard let entries = try? FileManager.default.contentsOfDirectory(
                at: dayURL,
                includingPropertiesForKeys: [.isDirectoryKey],
                options: [.skipsHiddenFiles]) else {
                continue
            }
            directories.append(contentsOf: entries.filter { entry in
                (try? entry.resourceValues(forKeys: [.isDirectoryKey]))?
                    .isDirectory == true
                    && entry.lastPathComponent.contains("session_")
            })
        }
        return directories
    }

    /// Day folder paths relative to the sessions root, newest first. The
    /// layout is `<year>/<month>/<day>`, so lexicographic ordering of the
    /// joined path matches chronological ordering.
    private func sortedDayFolderPaths() -> [String] {
        guard let yearEntries = try? FileManager.default.contentsOfDirectory(
            at: sessionsRootURL,
            includingPropertiesForKeys: nil,
            options: [.skipsHiddenFiles]) else {
            return []
        }
        var paths: [String] = []
        for yearURL in yearEntries {
            let monthRoot = sessionsRootURL
                .appendingPathComponent(yearURL.lastPathComponent)
            guard let monthEntries = try? FileManager.default
                .contentsOfDirectory(
                    at: monthRoot,
                    includingPropertiesForKeys: nil,
                    options: [.skipsHiddenFiles]) else {
                continue
            }
            for monthURL in monthEntries {
                let dayRoot = monthRoot
                    .appendingPathComponent(monthURL.lastPathComponent)
                guard let dayEntries = try? FileManager.default
                    .contentsOfDirectory(
                        at: dayRoot,
                        includingPropertiesForKeys: nil,
                        options: [.skipsHiddenFiles]) else {
                    continue
                }
                for dayURL in dayEntries {
                    paths.append(
                        "\(yearURL.lastPathComponent)/" +
                        "\(monthURL.lastPathComponent)/" +
                        dayURL.lastPathComponent)
                }
            }
        }
        return paths.sorted(by: >)
    }

    private func readSessionID(at sessionDirectory: URL) -> String? {
        let manifestURL = sessionDirectory
            .appendingPathComponent("manifest.json")
        guard let data = try? Data(contentsOf: manifestURL),
              let object = try? JSONSerialization.jsonObject(
                with: data) as? [String: Any],
              let sessionID = object["sessionId"] as? String,
              !sessionID.isEmpty else {
            return nil
        }
        return sessionID
    }

    // MARK: - Background tasks

    /// `bg_*` directories together with their own mtime, prefetched by the
    /// same listing call — thousands of finished tasks accumulate under the
    /// root, so no per-directory round trip is spent here.
    private func backgroundTaskDirectories() -> [(url: URL, modifiedAt: Date?)] {
        guard let entries = try? FileManager.default.contentsOfDirectory(
            at: backgroundTasksRootURL,
            includingPropertiesForKeys: [
                .isDirectoryKey,
                .contentModificationDateKey,
            ],
            options: [.skipsHiddenFiles]) else {
            return []
        }
        return entries.compactMap { entry in
            guard entry.lastPathComponent.hasPrefix("bg_"),
                  (try? entry.resourceValues(forKeys: [.isDirectoryKey]))?
                      .isDirectory == true else {
                return nil
            }
            let modifiedAt = (try? entry.resourceValues(
                forKeys: [.contentModificationDateKey]))?
                .contentModificationDate
            return (entry, modifiedAt)
        }
    }

    // MARK: - Shared helpers

    /// Newest modification time among a directory's immediate entries.
    /// In-place writes do not update the directory's own mtime, so the
    /// entries must be stat'd; the batch prefetch keeps this cheap.
    private func newestFileModification(in directory: URL) -> Date? {
        guard let entries = try? FileManager.default.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.contentModificationDateKey],
            options: [.skipsHiddenFiles]) else {
            return nil
        }
        return entries.compactMap { entry in
            (try? entry.resourceValues(
                forKeys: [.contentModificationDateKey]))?
                .contentModificationDate
        }
        .max()
    }

    /// Newest of the directory's own mtime (prefetched with the listing),
    /// `output.log`, and `summary.txt`. The two probe files are stat'd with
    /// plain string paths — building `URL`s and resolving per-URL resource
    /// values costs far more than the stats themselves once the root holds
    /// thousands of historical task folders.
    private func newestBackgroundTaskModification(
        in directory: URL,
        directoryModifiedAt: Date?
    ) -> Date? {
        var dates: [Date] = []
        if let directoryModifiedAt {
            dates.append(directoryModifiedAt)
        }
        let directoryPath = directory.path
        for fileName in ["output.log", "summary.txt"] {
            if let fileDate = Self.modificationDateIfExists(
                atPath: directoryPath + "/" + fileName) {
                dates.append(fileDate)
            }
        }
        return dates.max()
    }

    /// Direct POSIX stat: most historical task folders hold neither probe
    /// file, and `FileManager.attributesOfItem` pays an ObjC exception
    /// raise/catch for each miss — across thousands of folders that alone
    /// saturated a core. A raw `stat` call is a single microsecond syscall.
    private static func modificationDateIfExists(atPath path: String) -> Date? {
        var status = stat()
        guard stat(path, &status) == 0 else { return nil }
        return Date(
            timeIntervalSince1970: TimeInterval(status.st_mtimespec.tv_sec)
                + TimeInterval(status.st_mtimespec.tv_nsec) / 1_000_000_000)
    }
}

File: windows/src/AIQuotaBar.Providers/Minimax/MiniMaxActivityDetector.cs
// Swift 来源：AIQuotaBar/Services/MiniMaxActivityDetector.swift（整文件）。
// 对应测试：AIQuotaBar/Tests/Providers/MiniMaxActivityDetectorTests.swift。
//
// 检测两类本地运行时（只读元数据，永不读消息体 / 工具输出）：
// - MiniMax CLI 会话 ~/.minimax/v2/sessions/<y>/<m>/<d>/：活动 = 会话目录内最新文件 mtime
//   （manifest.json 只提供规范会话 ID —— 实测它可能落后实际写入数分钟）。
// - 后台任务 ~/.minimax/background-tasks/bg_<uuid>/：活动 = output.log / summary.txt / 任务目录 mtime。
// 环境变量 MINIMAX_HOME 可重定向根目录；目录路径注入式，默认值仅供生产装配。

using System.Text.Json;

namespace AIQuotaBar.Providers.Minimax;

/// <summary>MiniMax 本地活动检测器（Swift: MiniMaxActivityDetector）。</summary>
public sealed class MiniMaxActivityDetector
{
    /// <summary>Swift: defaultFreshnessWindow = 120s。</summary>
    public static readonly TimeSpan DefaultFreshnessWindow = TimeSpan.FromSeconds(120);

    /// <summary>
    /// 每轮扫描最近的几个日期目录（YYYY/MM/DD 名排序）。会话创建于启动当日，扫最新两个即可覆盖
    /// 跨午夜仍在工作的任务（Swift: scannedDayFolderLimit = 2）。
    /// </summary>
    public const int ScannedDayFolderLimit = 2;

    private static readonly string[] BackgroundTaskActivityFileNames = { "output.log", "summary.txt" };

    private readonly string _sessionsRootPath;
    private readonly string _backgroundTasksRootPath;
    private readonly TimeSpan _freshnessWindow;

    public MiniMaxActivityDetector(
        string? sessionsRootPath = null,
        string? backgroundTasksRootPath = null,
        TimeSpan? freshnessWindow = null)
    {
        _sessionsRootPath = sessionsRootPath ?? DefaultSessionsRootPath();
        _backgroundTasksRootPath = backgroundTasksRootPath ?? DefaultBackgroundTasksRootPath();
        _freshnessWindow = freshnessWindow ?? DefaultFreshnessWindow;
    }

    /// <summary>Swift: defaultHomeURL —— MINIMAX_HOME 非空优先，否则 ~/.minimax。</summary>
    public static string DefaultHomePath(IReadOnlyDictionary<string, string>? environment = null)
    {
        var configured = environment is not null && environment.TryGetValue("MINIMAX_HOME", out var overridePath)
            ? overridePath
            : Environment.GetEnvironmentVariable("MINIMAX_HOME");
        if (!string.IsNullOrEmpty(configured))
        {
            return configured;
        }
        return Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".minimax");
    }

    /// <summary>Swift: defaultSessionsRootURL —— &lt;home&gt;/v2/sessions。</summary>
    public static string DefaultSessionsRootPath(IReadOnlyDictionary<string, string>? environment = null) =>
        Path.Combine(DefaultHomePath(environment), "v2", "sessions");

    /// <summary>Swift: defaultBackgroundTasksRootURL —— &lt;home&gt;/background-tasks。</summary>
    public static string DefaultBackgroundTasksRootPath(IReadOnlyDictionary<string, string>? environment = null) =>
        Path.Combine(DefaultHomePath(environment), "background-tasks");

    /// <summary>Swift: isClientInstalled —— 任务保护资格的本地存在性检查。</summary>
    public static bool IsClientInstalled(IReadOnlyDictionary<string, string>? environment = null)
    {
        var home = DefaultHomePath(environment);
        return File.Exists(home) || Directory.Exists(home);
    }

    /// <summary>异步入口（Swift: snapshot() 在 detached task 上执行目录扫描）。</summary>
    public async Task<ProviderLocalActivitySnapshot> SnapshotAsync(CancellationToken cancellationToken = default) =>
        await Task.Run(() => DetectSnapshot(DateTimeOffset.UtcNow), cancellationToken).ConfigureAwait(false);

    /// <summary>同步检测核心（Swift: detectSnapshot(now:)），测试注入 now 与目录路径。</summary>
    public ProviderLocalActivitySnapshot DetectSnapshot(DateTimeOffset now)
    {
        var cutoff = now - _freshnessWindow;
        var activeSessionIds = new HashSet<string>(StringComparer.Ordinal);
        var lastEventBySession = new Dictionary<string, DateTimeOffset>(StringComparer.Ordinal);
        DateTimeOffset? lastEventAt = null;

        void Record(string prefixedSessionId, DateTimeOffset activityAt)
        {
            activeSessionIds.Add(prefixedSessionId);
            lastEventBySession[prefixedSessionId] = activityAt;
            if (lastEventAt is null || activityAt > lastEventAt)
            {
                lastEventAt = activityAt;
            }
        }

        foreach (var sessionDirectory in RecentSessionDirectories())
        {
            var activityAt = NewestFileModification(sessionDirectory);
            if (activityAt is null || activityAt < cutoff || activityAt > now)
            {
                continue;
            }
            var sessionId = ReadSessionId(sessionDirectory)
                ?? Path.GetFileName(sessionDirectory);
            Record($"minimax:cli:{sessionId}", activityAt.Value);
        }

        foreach (var taskDirectory in BackgroundTaskDirectories())
        {
            var activityAt = NewestBackgroundTaskModification(taskDirectory);
            if (activityAt is null || activityAt < cutoff || activityAt > now)
            {
                continue;
            }
            Record($"minimax:bg:{Path.GetFileName(taskDirectory)}", activityAt.Value);
        }

        return new ProviderLocalActivitySnapshot(activeSessionIds, lastEventAt, lastEventBySession);
    }

    // ------------------------------------------------------------------ CLI 会话

    private IReadOnlyList<string> RecentSessionDirectories()
    {
        var directories = new List<string>();
        foreach (var dayFolder in SortedDayFolderPaths().Take(ScannedDayFolderLimit))
        {
            var dayPath = Path.Combine(_sessionsRootPath, dayFolder);
            foreach (var entry in EnumerateNonHiddenEntries(dayPath))
            {
                if (!Directory.Exists(entry))
                {
                    continue;
                }
                if (!Path.GetFileName(entry).Contains("session_", StringComparison.Ordinal))
                {
                    continue;
                }
                directories.Add(entry);
            }
        }
        return directories;
    }

    /// <summary>
    /// 相对 sessions 根的日期目录路径，新→旧排序。布局是 &lt;year&gt;/&lt;month&gt;/&lt;day&gt;，
    /// 拼接路径的字典序与时间序一致（Swift: sortedDayFolderPaths）。
    /// </summary>
    private IReadOnlyList<string> SortedDayFolderPaths()
    {
        var paths = new List<string>();
        foreach (var year in EnumerateNonHiddenDirectories(_sessionsRootPath))
        {
            foreach (var month in EnumerateNonHiddenDirectories(year))
            {
                foreach (var day in EnumerateNonHiddenDirectories(month))
                {
                    paths.Add($"{Path.GetFileName(year)}/{Path.GetFileName(month)}/{Path.GetFileName(day)}");
                }
            }
        }
        return paths.OrderByDescending(path => path, StringComparer.Ordinal).ToList();
    }

    /// <summary>manifest.json 只用于规范会话 ID；缺失/损坏返回 null（回退目录名）。</summary>
    private string? ReadSessionId(string sessionDirectory)
    {
        var manifestPath = Path.Combine(sessionDirectory, "manifest.json");
        try
        {
            using var document = JsonDocument.Parse(File.ReadAllText(manifestPath));
            if (document.RootElement.ValueKind == JsonValueKind.Object
                && document.RootElement.TryGetProperty("sessionId", out var value)
                && value.ValueKind == JsonValueKind.String
                && value.GetString() is { Length: > 0 } sessionId)
            {
                return sessionId;
            }
        }
        catch (Exception ex) when (ex is IOException or JsonException or UnauthorizedAccessException)
        {
            // Swift: try? Data(contentsOf:) / try? JSONSerialization —— 静默回退。
        }
        return null;
    }

    // ------------------------------------------------------------------ 后台任务

    private IReadOnlyList<string> BackgroundTaskDirectories() =>
        EnumerateNonHiddenDirectories(_backgroundTasksRootPath)
            .Where(entry => Path.GetFileName(entry).StartsWith("bg_", StringComparison.Ordinal))
            .ToList();

    // ------------------------------------------------------------------ 共享辅助

    /// <summary>
    /// 目录直接子条目的最新 mtime。原位写不更新目录自身 mtime，必须逐条 stat（Swift 注释原意）。
    /// </summary>
    private DateTimeOffset? NewestFileModification(string directory)
    {
        DateTimeOffset? newest = null;
        foreach (var entry in EnumerateNonHiddenEntries(directory))
        {
            DateTimeOffset modified;
            try
            {
                var attributes = File.GetAttributes(entry);
                modified = (attributes & FileAttributes.Directory) == FileAttributes.Directory
                    ? new DateTimeOffset(Directory.GetLastWriteTimeUtc(entry))
                    : new DateTimeOffset(File.GetLastWriteTimeUtc(entry));
            }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
            {
                continue;
            }
            if (newest is null || modified > newest)
            {
                newest = modified;
            }
        }
        return newest;
    }

    /// <summary>后台任务目录自身 + output.log / summary.txt 的最新 mtime。</summary>
    private DateTimeOffset? NewestBackgroundTaskModification(string directory)
    {
        var dates = new List<DateTimeOffset>();
        try
        {
            dates.Add(new DateTimeOffset(Directory.GetLastWriteTimeUtc(directory)));
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            // 目录 stat 失败继续尝试文件。
        }
        foreach (var fileName in BackgroundTaskActivityFileNames)
        {
            var filePath = Path.Combine(directory, fileName);
            if (File.Exists(filePath))
            {
                dates.Add(new DateTimeOffset(File.GetLastWriteTimeUtc(filePath)));
            }
        }
        return dates.Count == 0 ? null : dates.Max();
    }

    /// <summary>跳过隐藏条目（Swift .skipsHiddenFiles：名字以 "." 开头）的目录枚举，失败返回空。</summary>
    private static IEnumerable<string> EnumerateNonHiddenDirectories(string path)
    {
        List<string> entries;
        try
        {
            // 枚举是惰性的，必须物化到 try 内才能捕获“目录不存在 / 无权限”。
            entries = Directory.EnumerateDirectories(path).ToList();
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            yield break;
        }
        foreach (var entry in entries)
        {
            if (!Path.GetFileName(entry).StartsWith(".", StringComparison.Ordinal))
            {
                yield return entry;
            }
        }
    }

    private static IEnumerable<string> EnumerateNonHiddenEntries(string path)
    {
        List<string> entries;
        try
        {
            entries = Directory.EnumerateFileSystemEntries(path).ToList();
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            yield break;
        }
        foreach (var entry in entries)
        {
            if (!Path.GetFileName(entry).StartsWith(".", StringComparison.Ordinal))
            {
                yield return entry;
            }
        }
    }
}
